File: broadcastreceiverlibrary/src/main/java/com/osapps/broadcastreceiverlibrary/network/NetworkReceiver.java
package com.osapps.broadcastreceiverlibrary.network;

import android.content.Intent;
import android.net.NetworkInfo;
import android.os.Bundle;
import android.support.annotation.NonNull;

import com.osapps.broadcastreceiverlibrary.generic.BroadcastReceiver;


/**
 * Created by osapps on 07/05/2018.
 */
public class NetworkReceiver extends BroadcastReceiver<NetworkObserver, NewNetworkStateObj> {


    @Override
    protected void notifyObserver(@NonNull NetworkObserver observer, @NonNull NewNetworkStateObj object) {
        observer.onNetworkChanged(object);
    }

    @NonNull
    @Override
    protected NewNetworkStateObj extractIntent(@NonNull Intent intent) {
        Bundle extras = intent.getExtras();
        assert extras != null;
        NetworkInfo info = extras.getParcelable("networkInfo");
        assert info != null;
        return new NewNetworkStateObj(info.getState());
    }

    @NonNull
    @Override
    protected String getIntentFilterName() {
        return "android.net.conn.CONNECTIVITY_CHANGE";
    }

}

File: broadcastreceiverlibrary/src/main/java/com/osapps/broadcastreceiverlibrary/network/NetworkReceiver.kt
package com.osapps.broadcastreceiverlibrary.network

import android.content.Intent
import android.net.NetworkInfo

import com.osapps.broadcastreceiverlibrary.generic.BroadcastReceiver


/**
 * Created by osapps on 07/05/2018.
 */

/**
 * This receiver meant to catch all of the network changes, like connected/disconnected and more.
 *
 * The return object will be a state, from [NetworkInfo.State] so use it to understand the state.
 *
 * To use, just inject the receiver and subscribe the observers to it.
 * And by god don't forget to add the network permissions in your app's manifest file.
 */
class NetworkReceiver : BroadcastReceiver<NetworkObserver, NewNetworkStateObj>() {

    override val intentFilterName: String
        get() = "android.net.conn.CONNECTIVITY_CHANGE"


    override fun notifyObserver(observer: NetworkObserver, broadcastObj: NewNetworkStateObj) {
        observer.onNetworkChanged(broadcastObj)
    }

    override fun extractIntent(intent: Intent): NewNetworkStateObj {
        val extras = intent.extras!!
        val info = extras.getParcelable<NetworkInfo>("networkInfo")!!
        return NewNetworkStateObj(info.state)
    }

}
